reference/src/taimpl/src/internal/buffer.c

```diff
@@ -41,25 +63,58 @@ sa_status convert_buffer(
             return status;
         }
 
-        svp_buffer_t* out_svp_buffer = svp_get_buffer(*svp);
-        *bytes = (uint8_t*) svp_get_svp_memory(out_svp_buffer) + buffer->context.svp.offset;
-        size_t out_length = svp_get_size(out_svp_buffer);
-        if (buffer->context.svp.offset + bytes_to_process > out_length) {
+        size_t memory_range;
+        if (add_overflow(buffer->context.svp.offset, bytes_to_process, &memory_range)) {
+            ERROR("Integer overflow");
+            return SA_STATUS_INVALID_PARAMETER;
+        }
+
+        svp_buffer_t* svp_buffer = svp_get_buffer(*svp);
+
+        // This call validates that SVP buffer is contained entirely within SVP memory.
+        void* memory_location = svp_get_svp_memory(svp_buffer);
+        if (memory_location == NULL) {
+            ERROR("memory range is not within SVP memory");
+            return SA_STATUS_INVALID_PARAMETER;
+        }
+
+        size_t memory_size = svp_get_size(svp_buffer);
+        if (memory_range > memory_size) {
             ERROR("buffer not large enough");
             return SA_STATUS_INVALID_PARAMETER;
         }
+
+        if (add_overflow((unsigned long) memory_location, buffer->context.svp.offset, (unsigned long*) bytes)) {
+            ERROR("Integer overflow");
+            return SA_STATUS_INVALID_PARAMETER;
+        }
     } else {
         if (buffer->context.clear.buffer == NULL) {
             ERROR("NULL buffer");
             return SA_STATUS_NULL_PARAMETER;
         }
 
-        if (buffer->context.clear.offset + bytes_to_process > buffer->context.clear.length) {
+        size_t memory_range;
+        if (add_overflow(buffer->context.clear.offset, bytes_to_process, &memory_range)) {
+            ERROR("Integer overflow");
+            return SA_STATUS_INVALID_PARAMETER;
+        }
+
+        if (memory_range > buffer->context.clear.length) {
             ERROR("buffer not large enough");
             return SA_STATUS_INVALID_PARAMETER;
         }
 
-        *bytes = (uint8_t*) buffer->context.clear.buffer + buffer->context.clear.offset;
+        if (!memory_is_valid_clear(buffer->context.clear.buffer, buffer->context.clear.length)) {
+            ERROR("memory location is not within SVP memory");
```

Comment:
Is this comment correct? This looks like it is detecting that some memory locations are within SVP memory.